File: test-construction-visual.js
// Visual testing script for construction system
// Run with: node test-construction-visual.js

const { chromium } = require('playwright');
const path = require('path');
const fs = require('fs');

(async () => {
  const browser = await chromium.launch({ 
    headless: false,
    args: ['--start-maximized']
  });
  
  const context = await browser.newContext({
    viewport: { width: 1920, height: 1080 },
    deviceScaleFactor: 1,
  });
  
  const page = await context.newPage();
  
  // Create screenshots directory
  const screenshotDir = path.join(__dirname, 'construction-screenshots');
  if (!fs.existsSync(screenshotDir)) {
    fs.mkdirSync(screenshotDir);
  }
  
  try {
    console.log('1. Loading game...');
    await page.goto('http://localhost:5173');
    await page.waitForLoadState('networkidle');
    await page.waitForTimeout(2000);
    
    // Take initial screenshot
    await page.screenshot({ 
      path: path.join(screenshotDir, '01-initial.png'),
      fullPage: true 
    });
    
    console.log('2. Adding test materials...');
    await page.keyboard.press('t');
    await page.waitForTimeout(500);
    
    console.log('3. Selecting construction tool...');
    // Open construction UI
    await page.evaluate(() => {
      const gameWindow = window as any;
      const game = gameWindow.game;
      const constructionSlot = game.unifiedInventorySystem.getSlots()
        .findIndex(slot => slot.item?.id === 'construction_tool');
      game.unifiedInventorySystem.setActiveHotbarSlot(constructionSlot);
    });
    
    await page.waitForTimeout(1000);
    await page.screenshot({ 
      path: path.join(screenshotDir, '02-construction-ui.png'),
      fullPage: true 
    });
    
    console.log('4. Testing foundation placement...');
    // Click foundation button
    await page.click('button:has-text("Foundation")');
    await page.waitForTimeout(500);
    
    // Move mouse to center and take screenshot of preview
    await page.mouse.move(960, 540);
    await page.waitForTimeout(500);
    await page.screenshot({ 
      path: path.join(screenshotDir, '03-foundation-preview.png'),
      fullPage: true 
    });
    
    // Place foundation
    await page.mouse.click(960, 540);
    await page.waitForTimeout(500);
    await page.screenshot({ 
      path: path.join(screenshotDir, '04-foundation-placed.png'),
      fullPage: true 
    });
    
    console.log('5. Testing wall placement...');
    await page.click('button:has-text("Wall")');
    await page.waitForTimeout(500);
    
    // Show wall preview
    await page.mouse.move(960, 540);
    await page.waitForTimeout(500);
    await page.screenshot({ 
      path: path.join(screenshotDir, '05-wall-preview.png'),
      fullPage: true 
    });
    
    // Place walls in a line
    await page.mouse.click(960, 540);
    await page.mouse.move(1100, 540);
    await page.waitForTimeout(500);
    await page.screenshot({ 
      path: path.join(screenshotDir, '06-wall-line-preview.png'),
      fullPage: true 
    });
    
    await page.mouse.click(1100, 540);
    await page.waitForTimeout(500);
    await page.screenshot({ 
      path: path.join(screenshotDir, '07-walls-placed.png'),
      fullPage: true 
    });
    
    console.log('6. Testing multi-level building...');
    // Place more foundations
    await page.click('button:has-text("Foundation")');
    for (let i = 0; i < 3; i++) {
      await page.mouse.click(820 + i * 140, 540);
      await page.waitForTimeout(200);
    }
    
    // Add walls
    await page.click('button:has-text("Wall")');
    await page.mouse.click(820, 540);
    await page.mouse.click(1100, 540);
    await page.waitForTimeout(500);
    
    // Go up a level
    await page.keyboard.press('PageUp');
    await page.waitForTimeout(500);
    await page.screenshot({ 
      path: path.join(screenshotDir, '08-level-1-view.png'),
      fullPage: true 
    });
    
    // Place floor
    await page.click('button:has-text("Floor")');
    await page.mouse.click(820, 540);
    await page.mouse.move(1100, 640);
    await page.waitForTimeout(500);
    await page.screenshot({ 
      path: path.join(screenshotDir, '09-floor-preview.png'),
      fullPage: true 
    });
    
    await page.mouse.click(1100, 640);
    await page.waitForTimeout(500);
    await page.screenshot({ 
      path: path.join(screenshotDir, '10-floor-placed.png'),
      fullPage: true 
    });
    
    console.log('7. Testing different camera angles...');
    // Rotate camera
    await page.mouse.down({ button: 'middle' });
    await page.mouse.move(1200, 540);
    await page.mouse.up({ button: 'middle' });
    await page.waitForTimeout(500);
    await page.screenshot({ 
      path: path.join(screenshotDir, '11-rotated-view.png'),
      fullPage: true 
    });
    
    // Overhead view
    await page.keyboard.press('Tab');
    await page.waitForTimeout(500);
    await page.screenshot({ 
      path: path.join(screenshotDir, '12-overhead-view.png'),
      fullPage: true 
    });
    
    console.log('8. Testing visual helpers...');
    await page.keyboard.press('Tab'); // Back to third person
    await page.keyboard.press('g'); // Toggle grid
    await page.waitForTimeout(500);
    await page.screenshot({ 
      path: path.join(screenshotDir, '13-with-grid.png'),
      fullPage: true 
    });
    
    // Test pillar placement to see vertex highlights
    await page.click('button:has-text("Pillar")');
    await page.mouse.move(960, 540);
    await page.waitForTimeout(500);
    await page.screenshot({ 
      path: path.join(screenshotDir, '14-pillar-vertex-highlight.png'),
      fullPage: true 
    });
    
    console.log('9. Final structure view...');
    // Place some pillars
    await page.mouse.click(960, 540);
    await page.mouse.click(820, 540);
    await page.mouse.click(1100, 540);
    
    // Exit construction mode
    await page.keyboard.press('Escape');
    await page.waitForTimeout(500);
    
    // Move camera back for final view
    await page.mouse.down({ button: 'middle' });
    await page.mouse.move(800, 400);
    await page.mouse.up({ button: 'middle' });
    
    // Zoom out
    await page.mouse.wheel(0, 100);
    await page.waitForTimeout(500);
    
    await page.screenshot({ 
      path: path.join(screenshotDir, '15-final-structure.png'),
      fullPage: true 
    });
    
    console.log(`\nScreenshots saved to: ${screenshotDir}`);
    console.log('\nPlease review the screenshots to identify visual issues.');
    
    // Keep browser open for manual inspection
    console.log('\nBrowser will stay open for manual inspection. Press Ctrl+C to close.');
    await page.waitForTimeout(60000);
    
  } catch (error) {
    console.error('Error during visual test:', error);
    await page.screenshot({ 
      path: path.join(screenshotDir, 'error-state.png'),
      fullPage: true 
    });
  }
  
  await browser.close();
})();

File: test_construction_visual.py
#!/usr/bin/env python3
"""
Visual testing script for construction system
Run with: python3 test_construction_visual.py
"""

import os
import time
import subprocess
from datetime import datetime

print("Construction System Visual Testing")
print("="*50)

# Create screenshot directory
screenshot_dir = f"construction-test-{datetime.now().strftime('%Y%m%d-%H%M%S')}"
os.makedirs(screenshot_dir, exist_ok=True)

print(f"\nScreenshots will be saved to: {screenshot_dir}/")
print("\nPlease perform the following steps in the game:")
print("\n1. Open the game at http://localhost:5173")
print("2. Press 'T' to add test materials")
print("3. Select the construction tool (hammer icon)")
print("4. The Construction UI should appear at the bottom")
print("\nTake screenshots at each step by pressing PrintScreen")
print("Save them to the directory mentioned above")

print("\n" + "="*50)
print("VISUAL CHECKLIST:")
print("="*50)

print("\n[ ] Construction UI Appearance:")
print("    - Are all buttons visible and properly styled?")
print("    - Is the text readable?")
print("    - Are the material colors correct?")
print("    - Does the mode indicator update correctly?")

print("\n[ ] Preview System:")
print("    - Does the green preview appear when hovering?")
print("    - Does it snap to the hex grid properly?")
print("    - Does it turn red when placement is invalid?")
print("    - Are the edge/vertex highlights visible for walls/pillars?")

print("\n[ ] Foundation Placement:")
print("    - Does it place with a single click?")
print("    - Is the size appropriate for the hex?")
print("    - Is the texture/material correct?")

print("\n[ ] Wall Placement:")
print("    - Does the line mode work (click start, click end)?")
print("    - Do walls align to hex edges properly?")
print("    - Is the wall height correct?")
print("    - Do the path previews show correctly?")

print("\n[ ] Floor Placement:")
print("    - Does the fill mode work (click corners)?")
print("    - Does it require support from walls/pillars?")
print("    - Is the floor thickness appropriate?")

print("\n[ ] Multi-level Building:")
print("    - Can you go up levels with PageUp?")
print("    - Does the level indicator update?")
print("    - Do upper floors require proper support?")

print("\n[ ] Visual Helpers:")
print("    - Does the grid toggle with 'G'?")
print("    - Are alignment guides visible?")
print("    - Is the start marker visible in multi-placement modes?")
print("    - Do highlights render above other geometry?")

print("\n[ ] Material System:")
print("    - Do different materials have distinct colors?")
print("    - Are material buttons highlighted when selected?")

print("\n[ ] Camera Interaction:")
print("    - Can you rotate camera without interfering with placement?")
print("    - Does overhead view (Tab) work properly?")
print("    - Does zoom work as expected?")

print("\n" + "="*50)
print("\nOnce you've taken screenshots, we can review them together.")
print("Please describe what looks wrong or unexpected.")

input("\nPress Enter when you're ready to start testing...")

# Open the game in browser
print("\nOpening game in browser...")
try:
    subprocess.run(["xdg-open", "http://localhost:5173"])  # Linux
except:
    try:
        subprocess.run(["open", "http://localhost:5173"])  # macOS
    except:
        try:
            subprocess.run(["start", "http://localhost:5173"], shell=True)  # Windows
        except:
            print("Please open http://localhost:5173 in your browser manually")

print("\nWaiting for you to complete the visual test...")
input("Press Enter when done to see analysis instructions...")

print("\n" + "="*50)
print("SCREENSHOT ANALYSIS GUIDE:")
print("="*50)
print("\nWhen reviewing screenshots, look for:")
print("\n1. Geometry Issues:")
print("   - Incorrect mesh sizes or proportions")
print("   - Z-fighting (flickering overlapping surfaces)")
print("   - Missing faces or inside-out geometry")
print("   - Incorrect positioning or alignment")

print("\n2. Material/Rendering Issues:")
print("   - Wrong colors or transparency")
print("   - Missing textures")
print("   - Incorrect shader properties")
print("   - Depth sorting problems")

print("\n3. UI/UX Issues:")
print("   - Overlapping or misaligned UI elements")
print("   - Incorrect state indicators")
print("   - Missing visual feedback")
print("   - Confusing or unclear controls")

print("\n4. Interaction Issues:")
print("   - Preview not following cursor")
print("   - Incorrect snapping behavior")
print("   - Placement happening in wrong location")
print("   - Multi-placement modes not working as expected")

print("\nPlease describe what you see that looks wrong!")